# DSO 510

# Homework 1

## Before submitting your homework, make sure that all your code works. Any code that throws an error will not be graded and will be assumed to be incorrect.

## Keep in mind that when we provide you with code, you will still need to run that code. So don't just skip over the code that we provide you.

## Part 1:

### 1a. In the code box below create three variables. Create a string named location which contains the following: "Los Angeles". Create an integer named my_integer which contains the following value: 10. Create a float named my_float which contains the follow value: 10.0.  Next to each variable you create add a comment indicating what you are doing.

In [1]:
location = "Los Angeles" ##create a string variable named location
my_integer = 10 ##create an integer variable named my_integer
my_float = 10.0 ##create a float variable named my_float

### 1b. Write an if function with the following criteria:  If the value of the variable location (which you created in the above code)  is equal to "Los Angeles" then print the following: "You are in Los Angeles". Otherwise, your if function should print the following: "You are not in Los Angeles"

In [2]:
if location == "Los Angeles":
    print("You are in Los Angeles")
else:
    print("You are not in Los Angeles")

You are in Los Angeles


### 1c. Write an if function with the following criteria: if the value of the variable location is equal to "Los Angeles" and my_float is equal to 5 then print the following: "The criteria was met" , else, if both criteria are not met, print the following: "The criteria was not met"

In [3]:
if location == "Los Angeles" and my_float == 5:
    print("The criteria was met")
else:
    print("The criteria was not met")

The criteria was not met


### 1d. Below you will find code which creates and populates a list called my_list.

In [4]:
# This code is given, do not change anything in this box.
my_list = ["USC","UCLA","UC Davis","UCSD"]

#### Print the first element in my_list. Remember that in Python the list index starts at 0. That is, your output should be USC

In [5]:
print(my_list[0])

USC


#### Print the last element in my_list. Note, that while you could just count the elements and index the last one, i.e.: my_list[3], you can also use negative indexes to start from the end. For example, my_list[-1] will give you the same result: UCSD.

In [6]:
print(my_list[3])
print(my_list[-1])

UCSD
UCSD


### 1e. Use the modulo (%) opoerator and print the remainder of 15 divided by 7.

In [7]:
print(15%7)

1


### 1f. Below you are given code that creates two string: string_1 and string_2. Create a variable called combined_strings that concatenates the two strings with one space in between. After you create combined_strings print it.

In [8]:
# This code is given.  Do not change anything in this box.
string_1 = "hello"
string_2 = "my name is"

In [9]:
listStr = [string_1, string_2]
combined_strings = ' '.join(listStr)
combined_strings

'hello my name is'

## Part 2 

### 2a. Using a for loop, print out the numbers 4,5,6,7,8,9

In [10]:
for i in range(4,10):
    print(i)

4
5
6
7
8
9


### 2b. Using a for loop, print out the numbers 4,6,8,10

In [11]:
for i in range(4,11,2):
    print(i)

4
6
8
10


### 2c. Using a while loop, print out the numbers 5,6,7,8,9

In [12]:
n = 5
while n <10:
    print(n)
    n = n+1

5
6
7
8
9


### 2d. In the below code, we have created a list of votes, named votes, we obtained from a survey. The survey asked people for yes or no responses. However, the person inputting the data sometimes used capital letters and sometimes used lowercase letters.  Determine the percentages of yes and no votes in the list and print out the percentage. Note, you should not manually count the yes and no votes, that is, if we were to change the values of the votes list, your code should still be able to calculate the percentage of yes votes in the new list.

In [13]:
# This code is given.  Do not change anything in this box.
votes = ["y","Y","N","y","N","n","y","y","N","Y"]

In [15]:
#method 1
N = len(votes)
yes = []
no = []
r1 = (0, N)
n = 0
while n < N:
    if votes[n] == "y" or votes[n] == "Y":
        yes.append(votes[n])
    else:
        no.append(votes[n])
    n = n+1
v_yes = len(yes)
v_no = len(no)
perc_yes = v_yes / N
perc_no = v_no / N
print("The percentage of yes votes is",perc_yes)
print("The percentage of no votes is",perc_no)

The percentage of yes votes is 0.6
The percentage of no votes is 0.4


In [16]:
#method 2
y = votes.count("y") + votes.count("Y")
n = votes.count("n")+votes.count("N")
perc_Yes = y/N
perc_No = n/N
print("The percentage of yes votes is",perc_Yes)
print("The percentage of no votes is",perc_No)

The percentage of yes votes is 0.6
The percentage of no votes is 0.4


### 2e. Create a python list with the following values: [10,55,66,77,88,99]  and name it list_2.  Once you have created the list do the following: 
- Print the third element (66) of the list. [recall, the first element is in cell 0]
- Print the sum of all the values in the list.
- Print the mean of all the values in the list.
- Compute the product of all the numbers (multiply them all together) and print the result.
#### Remember, when you are printing the sum, the mean, and the product, your code should still work if we change the values in list_2.

In [17]:
list_2 = [10,55,66,77,88,99]
#1
print(list_2[2])
#2
print(sum(list_2))
#3
print(sum(list_2)/len(list_2))
#4
p = 1
for i in list_2:
    p *= i
print(p)

66
395
65.83333333333333
24350911200


## Part 3

## Steven keeps a list of the people that he knows in several dictionaries. Each dictionary has the name of the person as the key, and the address as the value.  Below you will find code that creates these dictionaries.  Run that code. After, please write code for the steps listed below:
- Print out the names of all of Steven's friends.
- Now print all of Steven's freinds' addresses.
- Michael seems to have betrayed Steven, remove him from the friends dictionary.
- Steven needs to mail everyone for a party. Make a single dictionary containing all the contents and their addresses.
- Steven is visiting Barcelona and wants to visit Luis while he's there.  print Luis' address.
- Steven has moved to the USA and wants to cut ties with everyone. Remove everyone from the single dictionary you previously created containing all the contacts.
#### Please provide comments in your code so that it is clear what you are doing.

In [18]:
family = {'Jay': '123 Liverpool', 'Lisa': '567 Anfield'}
friends = dict(Michael='455 Newcastle', Didi='321 Munich')
colleagues = dict([('Luis','899 Barcelona'), ('Philippe', '332 Brazil')])

In [19]:
#Print out the names of all of Steven's friends.
#method1
for f in friends:
    print(f)

Michael
Didi


In [20]:
#method2
friends.keys()

dict_keys(['Michael', 'Didi'])

In [21]:
#Now print all of Steven's freinds' addresses.
for f in friends:
    print(friends[f])

455 Newcastle
321 Munich


In [22]:
#Michael seems to have betrayed Steven, remove him from the friends dictionary.
del friends['Michael']
friends

{'Didi': '321 Munich'}

In [23]:
#Steven needs to mail everyone for a party. Make a single dictionary containing all the contents and their addresses.
party = dict(family, **friends, **colleagues)
party

{'Jay': '123 Liverpool',
 'Lisa': '567 Anfield',
 'Didi': '321 Munich',
 'Luis': '899 Barcelona',
 'Philippe': '332 Brazil'}

In [24]:
for person in party:
    print(party[person])

123 Liverpool
567 Anfield
321 Munich
899 Barcelona
332 Brazil


In [25]:
#Steven is visiting Barcelona and wants to visit Luis while he's there. print Luis' address.
print(colleagues['Luis'])

899 Barcelona


In [26]:
#Steven has moved to the USA and wants to cut ties with everyone. Remove everyone from the single dictionary you previously created containing all the contacts.
party.clear()
party

{}

## Part 4

## Below we have created a list called distance. This list has dictionaries embedded inside and some of those dictionaries have other dictionaries embedded inside them. This type of response is typical when calling API's and it is useful to be comfortable with getting values from this sort of data. Below we will first walk you through a few items that we can get from the distance list followed by a few questions which you will be asked to answer.

In [27]:
distance = [{'distance': {'text': '248 mi', 'value': 398525},
  u'duration': {'text': '3 hours 59 mins', 'value': 14346},
  u'status': 'OK'},
 {'distance': {'text': '2,306 mi', 'value': 3711303},
  u'duration': {'text': '1 day 9 hours', 'value': 120030},
  u'status': 'OK'}]


In [28]:
distance

[{'distance': {'text': '248 mi', 'value': 398525},
  'duration': {'text': '3 hours 59 mins', 'value': 14346},
  'status': 'OK'},
 {'distance': {'text': '2,306 mi', 'value': 3711303},
  'duration': {'text': '1 day 9 hours', 'value': 120030},
  'status': 'OK'}]

#### Let's look at the first element in the list distance:

In [29]:
distance[0]

{'distance': {'text': '248 mi', 'value': 398525},
 'duration': {'text': '3 hours 59 mins', 'value': 14346},
 'status': 'OK'}

#### The first element is a dictionary. Let's look at the keys of that dictionary:

In [30]:
distance[0].keys()

dict_keys(['distance', 'duration', 'status'])

#### It has three keys, distance, duration, and status. Let's look at what inside the distance key:

In [31]:
distance[0]["distance"]

{'text': '248 mi', 'value': 398525}

#### Here we find another dictionary with two keys, text and value.  So if we want the value of the 'text' key we would use the following:

In [32]:
print(distance[0]["distance"]["text"])

248 mi


#### The distance list is actually a response from a call to the Google maps API. The first element in the list is the response for the distance between location x and location y. The second element in the list is the response for the distance between location x and location z. So, 248 mi, what we got above, is actually the number of miles between location x and location y. The duration to get from location x to location y is 3 hours 59 mins. 

#### 4a:  print the following: 
- The distance between location x and location z (using the second element in the distance list). 
- The status of the request for the distance between location x and location y.

In [33]:
print(distance[1]["distance"]["text"])
print(distance[0]["status"])

2,306 mi
OK


## Part 5: PANDAS

### In this section we will use the pandas library. So, first import the Pandas library:

In [34]:
#import pandas library here:
import pandas as pd

## Accompanying this HW is a file called, census_tract.csv.  This file contains a snippet of data that we downloaded from the US Census Bureau. We have modified the data headers and columns for usability. Follow the steps below:

#### 1) Use pandas, specifically the pandas read_csv function, to assign the content in the census_tract.csv file into a dataframe variable named census_tract:

In [35]:
# To get you started we provide this code for you:
census_tract = pd.read_csv("census_tract.csv")

#### Write code that shows the the first 5 lines of census_tract using the .head function:

In [36]:
census_tract.head()

GEO.id2           HC01_EST_VC02 HC03_EST_VC02 HC05_EST_VC02  \
0  census_tract  total_population_18_24          male        female   
1    6001400100                     177            48           129   
2    6001400200                      65            29            36   
3    6001400300                     250            79           171   
4    6001400400                     152            16           136   

          HC01_EST_VC03 HC01_EST_VC04 HC01_EST_VC05  
0  less_than_highschool    highschool  some_college  
1                    73            27            49  
2                     0             5            14  
3                     0            37           155  
4                     0            29            69

#### You will notice that the first row was automatically assigned as the column names.  We actually want the second row to be the column name and we want to skip the first row.  

#### Use the pandas read_csv funtion again, to assign the contents in the csv file census_tract.csv into a dataframe names census tract, except this time skip the first row.  

#### For help, look at the documentation for read_csv: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html .  Specifically, you can look at the header option.

In [37]:
census_tract = pd.read_csv("census_tract.csv", header = 1)
census_tract.head()

census_tract  total_population_18_24  male  female  less_than_highschool  \
0    6001400100                     177    48     129                    73   
1    6001400200                      65    29      36                     0   
2    6001400300                     250    79     171                     0   
3    6001400400                     152    16     136                     0   
4    6001400500                     372   138     234                    37   

   highschool  some_college  
0          27            49  
1           5            14  
2          37           155  
3          29            69  
4         142           142

#### Now, the columns are appropriately labeled.  This data contains, for each census tract in California, the number of people aged 18-24,  the number of males, females, and the number of people with different education levels. This is only for people aged 18-24. 

#### Add a column to the census_tract dataframe that has how many people have a college degree or higher. This would be the total_population_18_24 column minus the sum of the less_than_highschool, highschool, and some_college columns

In [38]:
census_tract["no_less_than_college degree"] = census_tract.apply(lambda x: x[1] - x[4] - x[5] - x[6], axis = 1)
census_tract

census_tract  total_population_18_24  male  female  \
0       6001400100                     177    48     129   
1       6001400200                      65    29      36   
2       6001400300                     250    79     171   
3       6001400400                     152    16     136   
4       6001400500                     372   138     234   
5       6001400600                     138    47      91   
6       6001400700                     284   128     156   
7       6001400800                     266    82     184   
8       6001400900                     188   130      58   
9       6001401000                     779   328     451   
10      6001401100                     323   122     201   
11      6001401200                     154    66      88   
12      6001401300                     223   186      37   
13      6001401400                     332   141     191   
14      6001401500                     218   121      97   
15      6001401600                     168    72      96   
16      6001401700                     142    64      78   
17      6001401800                     267   148     119   
18      6001402200                     282   149     133   
19      6001402400                     176    69     107   
20      6001402500                     195    80     115   
21      6001402600                      61    19      42   
22      6001402700                     300   132     168   
23      6001402800                     362   284      78   
24      6001402900                     119    37      82   
25      6001403000                     206   104     102   
26      6001403100                     138   122      16   
27      6001403300                     544   192     352   
28      6001403400                     159    32     127   
29      6001403501                     233    78     155   
...            ...                     ...   ...     ...   
8027    6113010704                     910   273     637   
8028    6113010800                     372   159     213   
8029    6113010901                     652   295     357   
8030    6113010902                     585   274     311   
8031    6113011001                     696   239     457   
8032    6113011002                     322   123     199   
8033    6113011101                     420   234     186   
8034    6113011102                     596   227     369   
8035    6113011103                     313   197     116   
8036    6113011203                     125    58      67   
8037    6113011204                     518   192     326   
8038    6113011205                     766   352     414   
8039    6113011206                     742   217     525   
8040    6113011300                     783   377     406   
8041    6113011400                     435   160     275   
8042    6113011500                     616   336     280   
8043    6115040100                     631   405     226   
8044    6115040200                    1095   587     508   
8045    6115040301                     333   117     216   
8046    6115040302                     271   112     159   
8047    6115040303                     685   364     321   
8048    6115040400                     627   295     332   
8049    6115040500                     406   212     194   
8050    6115040600                     402   273     129   
8051    6115040700                     927   497     430   
8052    6115040800                     514   264     250   
8053    6115040901                     425   176     249   
8054    6115040902                     642   416     226   
8055    6115041000                     298    88     210   
8056    6115041100                     141   115      26   

      less_than_highschool  highschool  some_college  \
0                       73          27            49   
1                        0           5            14   
2                        0          37           155   
3                        0          29            69   
4    

#### The census tract column is a unique identifier for each census tract. What type of variable is it (string, int, etc.).  you can use the dtype function.  Print the type.

In [39]:
census_tract.dtypes

census_tract                   int64
total_population_18_24         int64
male                           int64
female                         int64
less_than_highschool           int64
highschool                     int64
some_college                   int64
no_less_than_college degree    int64
dtype: object

#### Let's examine the census tract id in more detail.  print the census tract id of the first row:

In [40]:
print(census_tract["census_tract"][0])

6001400100


#### It turns out the id is split into different components that provide details about the census tract. For example, the first number, 6, means that this census tract is in California. You will notice that all the census tracts start with 6. The next three digits in the census tract (001 for the first row) represent the county that the census tract is in. 

#### Create a new column, called county_id, which is an integer that has the value of those three digits. For example, county_id in row 1 is equal to 1 (since 001 is the integer 1) and for the very last row it is 115. 

In [41]:
def three_digit(x):
    temp = str(x)
    output = int(temp[1:4])
    return output

In [42]:
census_tract["county_id"] = census_tract["census_tract"].apply(three_digit)
census_tract

census_tract  total_population_18_24  male  female  \
0       6001400100                     177    48     129   
1       6001400200                      65    29      36   
2       6001400300                     250    79     171   
3       6001400400                     152    16     136   
4       6001400500                     372   138     234   
5       6001400600                     138    47      91   
6       6001400700                     284   128     156   
7       6001400800                     266    82     184   
8       6001400900                     188   130      58   
9       6001401000                     779   328     451   
10      6001401100                     323   122     201   
11      6001401200                     154    66      88   
12      6001401300                     223   186      37   
13      6001401400                     332   141     191   
14      6001401500                     218   121      97   
15      6001401600                     168    72      96   
16      6001401700                     142    64      78   
17      6001401800                     267   148     119   
18      6001402200                     282   149     133   
19      6001402400                     176    69     107   
20      6001402500                     195    80     115   
21      6001402600                      61    19      42   
22      6001402700                     300   132     168   
23      6001402800                     362   284      78   
24      6001402900                     119    37      82   
25      6001403000                     206   104     102   
26      6001403100                     138   122      16   
27      6001403300                     544   192     352   
28      6001403400                     159    32     127   
29      6001403501                     233    78     155   
...            ...                     ...   ...     ...   
8027    6113010704                     910   273     637   
8028    6113010800                     372   159     213   
8029    6113010901                     652   295     357   
8030    6113010902                     585   274     311   
8031    6113011001                     696   239     457   
8032    6113011002                     322   123     199   
8033    6113011101                     420   234     186   
8034    6113011102                     596   227     369   
8035    6113011103                     313   197     116   
8036    6113011203                     125    58      67   
8037    6113011204                     518   192     326   
8038    6113011205                     766   352     414   
8039    6113011206                     742   217     525   
8040    6113011300                     783   377     406   
8041    6113011400                     435   160     275   
8042    6113011500                     616   336     280   
8043    6115040100                     631   405     226   
8044    6115040200                    1095   587     508   
8045    6115040301                     333   117     216   
8046    6115040302                     271   112     159   
8047    6115040303                     685   364     321   
8048    6115040400                     627   295     332   
8049    6115040500                     406   212     194   
8050    6115040600                     402   273     129   
8051    6115040700                     927   497     430   
8052    6115040800                     514   264     250   
8053    6115040901                     425   176     249   
8054    6115040902                     642   416     226   
8055    6115041000                     298    88     210   
8056    6115041100                     141   115      26   

      less_than_highschool  highschool  some_college  \
0                       73          27            49   
1                        0           5            14   
2                        0          37           155   
3                        0          29            69   
4    

#### You also have another csv file called county_data.csv . In this file you have two columns.  The first column, called county_id, has the county id.  The second column has the name of the county.  First, use .read_csv and pandas to assign this csv file to a variable called county_data

In [166]:
county_data = pd.read_csv("county_data.csv", header = 0)
county_data

county_id         county_name
0          1      Alameda County
1          3       Alpine County
2         29         Kern County
3         31        Kings County
4         37  Los Angeles County
5         39       Madera County
6         55         Napa County
7         59       Orange County
8         67   Sacramento County
9        115         Yuba County

In [168]:
county_data.dtypes

county_id       int64
county_name    object
dtype: object

In [ ]:
county_data[county]

#### Merge the census_tract dataframe and the county_data dataframe and assign the outcome to a new dataframe called final_DF. Look at the documentation for merge: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html . 

#### If there is a row in the census_tract dataframe that does not have a match in the county_df then that row should be empty. If there is a row in the county_df that has a census_tract id that is not in the census_tract dataframe that should be ignored.

In [175]:
final_DF = pd.merge(census_tract, county_data, on = ['county_id'], how = 'left')
final_DF

census_tract  total_population_18_24  male  female  \
0       6001400100                     177    48     129   
1       6001400200                      65    29      36   
2       6001400300                     250    79     171   
3       6001400400                     152    16     136   
4       6001400500                     372   138     234   
5       6001400600                     138    47      91   
6       6001400700                     284   128     156   
7       6001400800                     266    82     184   
8       6001400900                     188   130      58   
9       6001401000                     779   328     451   
10      6001401100                     323   122     201   
11      6001401200                     154    66      88   
12      6001401300                     223   186      37   
13      6001401400                     332   141     191   
14      6001401500                     218   121      97   
15      6001401600                     168    72      96   
16      6001401700                     142    64      78   
17      6001401800                     267   148     119   
18      6001402200                     282   149     133   
19      6001402400                     176    69     107   
20      6001402500                     195    80     115   
21      6001402600                      61    19      42   
22      6001402700                     300   132     168   
23      6001402800                     362   284      78   
24      6001402900                     119    37      82   
25      6001403000                     206   104     102   
26      6001403100                     138   122      16   
27      6001403300                     544   192     352   
28      6001403400                     159    32     127   
29      6001403501                     233    78     155   
...            ...                     ...   ...     ...   
8027    6113010704                     910   273     637   
8028    6113010800                     372   159     213   
8029    6113010901                     652   295     357   
8030    6113010902                     585   274     311   
8031    6113011001                     696   239     457   
8032    6113011002                     322   123     199   
8033    6113011101                     420   234     186   
8034    6113011102                     596   227     369   
8035    6113011103                     313   197     116   
8036    6113011203                     125    58      67   
8037    6113011204                     518   192     326   
8038    6113011205                     766   352     414   
8039    6113011206                     742   217     525   
8040    6113011300                     783   377     406   
8041    6113011400                     435   160     275   
8042    6113011500                     616   336     280   
8043    6115040100                     631   405     226   
8044    6115040200                    1095   587     508   
8045    6115040301                     333   117     216   
8046    6115040302                     271   112     159   
8047    6115040303                     685   364     321   
8048    6115040400                     627   295     332   
8049    6115040500                     406   212     194   
8050    6115040600                     402   273     129   
8051    6115040700                     927   497     430   
8052    6115040800                     514   264     250   
8053    6115040901                     425   176     249   
8054    6115040902                     642   416     226   
8055    6115041000                     298    88     210   
8056    6115041100                     141   115      26   

      less_than_highschool  highschool  some_college  \
0                       73          27            49   
1                        0           5            14   
2                        0          37           155   
3                        0          29            69   
4    

#### You should have a column in the final_DF called county_name.  Some rows will have this row empty as it wasn't able to find a match in the county_data dataframe. 

#### Use the fillna function in Python to replace all the empty values with a string: "NA".  Here is the documentation for .fillna :  https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html . 

#### Think about the inplace option and whether you need it. Remember, you are replacing the final_DF and are not creating a new dataframe.

In [180]:
final_DF.fillna("NA", inplace=True)
final_DF

census_tract  total_population_18_24  male  female  \
0       6001400100                     177    48     129   
1       6001400200                      65    29      36   
2       6001400300                     250    79     171   
3       6001400400                     152    16     136   
4       6001400500                     372   138     234   
5       6001400600                     138    47      91   
6       6001400700                     284   128     156   
7       6001400800                     266    82     184   
8       6001400900                     188   130      58   
9       6001401000                     779   328     451   
10      6001401100                     323   122     201   
11      6001401200                     154    66      88   
12      6001401300                     223   186      37   
13      6001401400                     332   141     191   
14      6001401500                     218   121      97   
15      6001401600                     168    72      96   
16      6001401700                     142    64      78   
17      6001401800                     267   148     119   
18      6001402200                     282   149     133   
19      6001402400                     176    69     107   
20      6001402500                     195    80     115   
21      6001402600                      61    19      42   
22      6001402700                     300   132     168   
23      6001402800                     362   284      78   
24      6001402900                     119    37      82   
25      6001403000                     206   104     102   
26      6001403100                     138   122      16   
27      6001403300                     544   192     352   
28      6001403400                     159    32     127   
29      6001403501                     233    78     155   
...            ...                     ...   ...     ...   
8027    6113010704                     910   273     637   
8028    6113010800                     372   159     213   
8029    6113010901                     652   295     357   
8030    6113010902                     585   274     311   
8031    6113011001                     696   239     457   
8032    6113011002                     322   123     199   
8033    6113011101                     420   234     186   
8034    6113011102                     596   227     369   
8035    6113011103                     313   197     116   
8036    6113011203                     125    58      67   
8037    6113011204                     518   192     326   
8038    6113011205                     766   352     414   
8039    6113011206                     742   217     525   
8040    6113011300                     783   377     406   
8041    6113011400                     435   160     275   
8042    6113011500                     616   336     280   
8043    6115040100                     631   405     226   
8044    6115040200                    1095   587     508   
8045    6115040301                     333   117     216   
8046    6115040302                     271   112     159   
8047    6115040303                     685   364     321   
8048    6115040400                     627   295     332   
8049    6115040500                     406   212     194   
8050    6115040600                     402   273     129   
8051    6115040700                     927   497     430   
8052    6115040800                     514   264     250   
8053    6115040901                     425   176     249   
8054    6115040902                     642   416     226   
8055    6115041000                     298    88     210   
8056    6115041100                     141   115      26   

      less_than_highschool  highschool  some_college  \
0                       73          27            49   
1                        0           5            14   
2                        0          37           155   
3                        0          29            69   
4    

In [181]:
final_DF["county_name"].isna().sum()

0